File: reference/fixtures/ConstantVelocity.cpp
// Reference-only open-loop velocity driver; this is not an upstream SCRIMMAGE plugin.
// No retained upstream autonomy writes a nonzero velocity command, so this
// supplies the same constant ENU command in both languages to test the
// UNMODIFIED upstream SingleIntegratorControllerSimple and SingleIntegrator.
#include <cmath>
#include <sstream>
#include <stdexcept>
#include <vector>

#include <scrimmage/autonomy/Autonomy.h>
#include <scrimmage/common/VariableIO.h>
#include <scrimmage/plugin_manager/RegisterPlugin.h>

class ConstantVelocity : public scrimmage::Autonomy {
 public:
    void init(std::map<std::string, std::string>& params) override {
        std::istringstream input(params.at("velocity"));
        double value;
        while (input >> value) {
            if (!std::isfinite(value)) throw std::runtime_error("nonfinite velocity");
            velocity_.push_back(value);
        }
        if (velocity_.size() != 3 || !input.eof()) {
            throw std::runtime_error("ConstantVelocity.velocity needs three numbers");
        }
        using Type = scrimmage::VariableIO::Type;
        using Direction = scrimmage::VariableIO::Direction;
        ports_ = {vars_.declare(Type::velocity_x, Direction::Out),
                  vars_.declare(Type::velocity_y, Direction::Out),
                  vars_.declare(Type::velocity_z, Direction::Out)};
    }

    bool step_autonomy(double, double) override {
        for (std::size_t i = 0; i < 3; ++i) vars_.output(ports_[i], velocity_[i]);
        return true;
    }

 private:
    std::vector<int> ports_;
    std::vector<double> velocity_;
};

REGISTER_PLUGIN(scrimmage::Autonomy, ConstantVelocity, ConstantVelocity_plugin)

File: crates/core/src/plugin/autonomy/constant_velocity/constant_velocity.rs
//! Rust-only open-loop velocity test driver, not a navigation behavior.
//! Autonomy phase: write a configured constant ENU velocity to velocity_x/y/z.
//! The reference fixture supplies the same commands to the upstream C++ models.

use anyhow::{Result, ensure};
use serde::Deserialize;

use crate::plugin::{
    AgentContext, Autonomy, Frame, Plugin, PluginIo, PluginParams, Port, Ports, Unit, Update,
};

const PORTS: [&str; 3] = ["velocity_x", "velocity_y", "velocity_z"];

/// Mission parameters; `Default` supplies any key the mission leaves out.
#[derive(Deserialize)]
#[serde(default, deny_unknown_fields)]
struct ConstantVelocityParams {
    /// ENU velocity in m/s, written as three numbers: `"x y z"`.
    #[serde(rename = "velocity")]
    velocity_world_mps: Vec<f64>,
}

impl Default for ConstantVelocityParams {
    fn default() -> Self {
        Self {
            velocity_world_mps: vec![0.0; 3],
        }
    }
}

pub struct ConstantVelocity {
    velocity_world_mps: [f64; 3],
}

impl Plugin for ConstantVelocity {
    type Config = [f64; 3];

    fn configure(params: &PluginParams<'_>) -> Result<[f64; 3]> {
        let params: ConstantVelocityParams = params.parse()?;
        ensure!(
            params.velocity_world_mps.len() == 3,
            "ConstantVelocity.velocity needs three numbers"
        );
        Ok([
            params.velocity_world_mps[0],
            params.velocity_world_mps[1],
            params.velocity_world_mps[2],
        ])
    }

    fn new(velocity_world_mps: &[f64; 3]) -> Self {
        Self {
            velocity_world_mps: *velocity_world_mps,
        }
    }

    fn ports(_: &[f64; 3]) -> Ports {
        let mut ports = Ports::default();
        for name in PORTS {
            ports = ports.output(Port::new(name, Unit::MetersPerSecond, Frame::World));
        }
        ports
    }
}

impl Autonomy for ConstantVelocity {
    fn step(&mut self, _: &mut AgentContext<'_>, io: &mut PluginIo) -> Result<Update> {
        for (name, value) in PORTS.into_iter().zip(self.velocity_world_mps) {
            io.write(name, value)?;
        }
        Ok(Update::Applied)
    }
}
